source/h1_encoder.c

```diff
@@ -319,6 +343,347 @@ int aws_h1_encoder_start_message(
+            /* Ensure we're not at end-of-stream too early */
+            struct aws_stream_status status;
+            err = aws_input_stream_get_status(encoder->message->body, &status);
+            if (err) {
+                ENCODER_LOGF(
+                    TRACE,
+                    encoder,
+                    "Failed to query body stream status, error %d (%s)",
+                    aws_last_error(),
+                    aws_error_name(aws_last_error()));
+
+                return AWS_OP_ERR;
+            }
+            if (status.is_end_of_stream) {
+                ENCODER_LOGF(
+                    ERROR,
+                    encoder,
+                    "Reached end of body stream before Content-Length: %" PRIu64 " sent",
+                    encoder->message->content_length);
+                return aws_raise_error(AWS_ERROR_HTTP_OUTGOING_STREAM_LENGTH_INCORRECT);
+            }
+
+            ENCODER_LOG(
+                TRACE,
+                encoder,
+                "No body data written, concluding this message. "
+                "Will try to write body data again in the next message.");
+            return AWS_OP_SUCCESS;
+        }
+    }
+    return AWS_OP_SUCCESS;
+}
+
+static bool s_write_crlf(struct aws_byte_buf *dst) {
+    char crlf[2] = {'\r', '\n'};
+    struct aws_byte_cursor crlf_cursor = aws_byte_cursor_from_array(&crlf, AWS_ARRAY_SIZE(crlf));
+    return aws_byte_buf_write_from_whole_cursor(dst, crlf_cursor);
+}
+
+static bool s_chunk_size_state(struct aws_byte_buf *dst, struct aws_h1_encoder *encoder, int *aws_op_result) {
+    AWS_ASSERT(dst);
+    AWS_ASSERT(encoder);
+    AWS_ASSERT(aws_op_result);
+    int64_t chunk_size = 0;
+    if (aws_input_stream_get_length(encoder->message->body_chunk->stream, &chunk_size)) {
+        ENCODER_LOGF(
+            TRACE,
+            encoder,
+            "Failed to get length of body stream, error %d (%s)",
+            aws_last_error(),
+            aws_error_name(aws_last_error()));
+        *aws_op_result = AWS_OP_ERR;
+        return false;
+    }
+    char ascii_hex_chunk_size_str[sizeof(size_t) * 2 + 1] = {0};
+    int ascii_hex_len = snprintf(ascii_hex_chunk_size_str, sizeof(ascii_hex_chunk_size_str), "%zX", chunk_size);
+    if ((dst->capacity - dst->len) <= (size_t)ascii_hex_len) {
+        return false;
+    }
+    return aws_byte_buf_write_from_whole_cursor(dst, aws_byte_cursor_from_c_str(ascii_hex_chunk_size_str));
+}
+
+static bool s_fill_byte_buffer(struct aws_byte_buf *dst, struct aws_byte_cursor *src) {
+    size_t write_size = aws_min_size(dst->capacity - dst->len, src->len);
+    struct aws_byte_cursor sub_cursor;
+    sub_cursor.len = write_size;
+    sub_cursor.ptr = src->ptr;
+    if (AWS_UNLIKELY(AWS_OP_SUCCESS != aws_byte_buf_append(dst, &sub_cursor))) {
+        return false;
+    }
+    aws_byte_cursor_advance(src, write_size);
+    return 0 == src->len;
+}
+
+#define fall_through __attribute__((fallthrough))
```

Comment:
quite likely, gcc and clang only, might have to move this to aws-c-common's platform macros.

Reply:
How would you feel about a goto label instead of the fallthrough for a portable solution? 

e.g.
...
switch(integer) {
    case 1:
       /*... stuff */
      goto two;
    case 2:
    two:
      /* ... more stuff */
    default:
        break;
}
...